File: tools/mstand/squeeze_lib/lib/squeezer_web_surveys.cpp
#include "squeezer_web_surveys.h"

#include <quality/ab_testing/lib_calc_session_metrics/common/words.h> // NSessionsMetrics::GetQuery
#include <tools/mstand/squeeze_lib/requests/common/base_request.h> // NMstand::GetSerpType

namespace NMstand {

using namespace NYT;

/////////////////////////////// TWebSurveysActionsSqueezer
TTableSchema TWebSurveysActionsSqueezer::GetSchema() const {
    TTableSchema schema;
    schema.AddColumn(TColumnSchema().Name("yuid").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("ts").Type(NYT::VT_INT64));
    schema.AddColumn(TColumnSchema().Name("action_index").Type(NYT::VT_INT64));
    schema.AddColumn(TColumnSchema().Name("bucket").Type(NYT::VT_INT64));
    schema.AddColumn(TColumnSchema().Name("is_match").Type(NYT::VT_BOOLEAN));
    schema.AddColumn(TColumnSchema().Name("testid").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("servicetype").Type(NYT::VT_STRING));

    schema.AddColumn(TColumnSchema().Name("answer_id").Type(NYT::VT_INT32));
    schema.AddColumn(TColumnSchema().Name("answers").Type(NYT::VT_ANY));
    schema.AddColumn(TColumnSchema().Name("block_id").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("block_name").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("browser").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("comment_text").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("event_id").Type(NYT::VT_STRING));
    schema.AddColumn("multi_answers", NTi::Optional(NTi::List(NTi::Tuple({{NTi::Int32()}, {NTi::List(NTi::Int32())}}))));
    schema.AddColumn(TColumnSchema().Name("query_text").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("question_id").Type(NYT::VT_INT32));
    schema.AddColumn(TColumnSchema().Name("questions").Type(NYT::VT_ANY));
    schema.AddColumn(TColumnSchema().Name("reqid").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("result_block_id").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("selected_option").Type(NYT::VT_INT32));
    schema.AddColumn(TColumnSchema().Name("survey_id").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("survey_type").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("type").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("ui").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("url").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("visibility").Type(NYT::VT_BOOLEAN));
    schema.AddColumn(TColumnSchema().Name("wizard_name").Type(NYT::VT_STRING));
    schema.AddColumn(TColumnSchema().Name("prism_segment").Type(NYT::VT_UINT64));
    schema.AddColumn(TColumnSchema().Name("dwelltime").Type(NYT::VT_UINT64));
    schema.AddColumn(TColumnSchema().Name("wizard_position").Type(NYT::VT_STRING));
    
    return schema.Strict(true);
}

ui32 TWebSurveysActionsSqueezer::GetVersion() const
{
    return 1;
}

bool TWebSurveysActionsSqueezer::CheckRequest(const NRA::TRequest* request) const
{
    return dynamic_cast<const NRA::TTouchYandexWebRequest*>(request) != nullptr
        || dynamic_cast<const NRA::TPadYandexWebRequest*>(request) != nullptr
        || dynamic_cast<const NRA::TMobileAppYandexWebRequest*>(request) != nullptr
        || dynamic_cast<const NRA::TYandexWebRequest*>(request) != nullptr;
}

// In order to convert NBaobab::TAttr -> NYT::TNode
namespace {
    NYT::TNode SerializeToNode(const NBaobab::TAttrsMap&);
    NYT::TNode SerializeToNode(const NBaobab::TAttrsArray&);

    NYT::TNode SerializeToNode(const NBaobab::TAttr& attr) {
        switch (attr.GetType()) {
            case NBaobab::EAttrType::Double:
                return attr.Get<double>();
            case NBaobab::EAttrType::Int:
                return attr.Get<i64>();
            case NBaobab::EAttrType::UnsignedInt:
                return attr.Get<ui64>();
            case NBaobab::EAttrType::String:
                return attr.Get<TString>();
            case NBaobab::EAttrType::Bool:
                return attr.Get<bool>();
            case NBaobab::EAttrType::Array:
                return SerializeToNode(attr.Get<NBaobab::TAttrsArray>());
            case NBaobab::EAttrType::Dict:
                return SerializeToNode(attr.Get<NBaobab::TAttrsMap>());
            default:
                return NYT::TNode::CreateEntity();
        }
    }

    NYT::TNode SerializeToNode(const NBaobab::TAttrsMap& attrsMap) {
        auto res = NYT::TNode::CreateMap();
        for (const auto& [k,v] : attrsMap) {
            res[k] = SerializeToNode(v);
        }
        return res;
    }

    NYT::TNode SerializeToNode(const NBaobab::TAttrsArray& attrsArray) {
        auto res = NYT::TNode::CreateList();
        for (const auto& el: attrsArray) {
            auto val = SerializeToNode(el);
            res.Add(val);
        }
        return res;
    }
}

namespace {
    TMaybe<i64> GetAnswerId(const NBaobab::TBlock& answerBlock) {
        if (const auto& id = NBaobab::GetAttrValueUnsafe<i64>(answerBlock, "answer_id")) {
            return id.GetRef();
        }
        if (const auto& id = NBaobab::GetAttrValueUnsafe<i64>(answerBlock, "id")) {
            return id.GetRef();
        }
        return Nothing();
    }

    NYT::TNode GetAnswers(const NBaobab::TBlock& surveyBlock) {
        NYT::TNode::TListType answers;
        for (const auto& answerBlock : NBaobab::TBFSer(surveyBlock,  NBaobab::NameFilter("answer"))) {
            if (const auto& answerId = GetAnswerId(answerBlock); answerId.Defined()) {
                answers.push_back(answerId.GetRef());
            }
        }
        return NYT::TNode::CreateList(answers);
    }

    NYT::TNode GetMultiAnswers(const NBaobab::TBlock& surveyBlock) {
        THashMap<i32, NYT::TNode::TListType> answers;
        for (const auto& answerBlock : NBaobab::TBFSer(surveyBlock,  NBaobab::NameFilter("answer"))) {
            if (const auto& idQuestion = NBaobab::GetAttrValueUnsafe<i64>(answerBlock, "idQuestion")) {
                if (const auto& answerId = GetAnswerId(answerBlock); answerId.Defined()) {
                    answers[idQuestion.GetRef()].push_back(answerId.GetRef());
                }
            }
        }
        NYT::TNode::TListType result;
        for (const auto& [k, v] : answers) {
            result.push_back(
                NYT::TNode::CreateList({NYT::TNode(k), NYT::TNode::CreateList(v)})
            );
        }
        return NYT::TNode::CreateList(result);
    }

    TString GetSurveyType(const NBaobab::TBlock& surveyBlock) {
        for (const auto& answerBlock : NBaobab::TBFSer(surveyBlock,  NBaobab::NameFilter("answer"))) {
            if (answerBlock.GetAttrs().has("text")) {
                return "text";
            }
            if (answerBlock.GetAttrs().has("emoji")) {
                return "emoji";
            }
        }
        return "unknown";
    }

    TMaybe<bool> CheckBlockVisible(
        const NRA::TRequest* request,
        const NBaobab::TRequestJoiner& joiner,
        const NBaobab::TBlock& block
    ) {
        for (const auto& event : joiner.GetEventsByBlockSubtree(block)) {
            if (const auto click = dynamic_cast<const NBaobab::IClick*>(event.Get())) {
                return true;
            }
        }

        NMouseTrack::TMouseTrackRequestData mouseTrackRequestData = NMouseTrack::ExtractFromRequest(*request);
        for (const auto& mouseTrackBlock : mouseTrackRequestData.BlocksData) {
            if (mouseTrackBlock.Get()->HasBlockCounterId() && block.GetID() == mouseTrackBlock.Get()->GetBlockCounterId()) {
                auto blockData = mouseTrackBlock.Get();
                if (blockData && blockData->HasTop()) {
                    double mouseTrackTop = blockData->GetTop();
                    if (mouseTrackTop > 0 && mouseTrackTop < 20000) {
                        return mouseTrackRequestData.MaxScrollYBottom > mouseTrackTop;
                    }
                }
            }
        }
        return Nothing();
    }

    TMaybe<int64_t> GetPrismSegment(const NRA::TRequest* request) {
        const auto* misc = dynamic_cast<const NRA::TMiscRequestProperties*>(request);
        if (misc) {
            const NRA::TNameValueMap& searchPropsUpper = misc->GetSearchPropsValues();
            if (const auto* valuePtr = searchPropsUpper.FindPtr("UPPER.PrismBigBLog.prism_segment")) {
                int64_t value;
                if (TryFromString(*valuePtr, value)) {
                    return value;
                }                
            }
        }
        return Nothing();
    }

    NYT::TNode GetRequestData(
        const NRA::TRequest* request,
        const NBaobab::TRequestJoiner& joiner,
        const NBaobab::TBlock& surveyBlock
    ) {
        auto requestData = NYT::TNode::CreateMap();
        requestData["yuid"] = request->GetUID();
        requestData["ts"] = request->GetTimestamp();
        requestData["reqid"] = request->GetReqID();
        requestData["type"] = "request";
        requestData["browser"] = request->GetBrowser();
        requestData["ui"] = ToString(GetSerpType(*request));
        if (const auto& queryText = NSessionsMetrics::GetQuery(request)) {
            requestData["query_text"] = queryText.GetRef();
        }

        if (const auto& resultBlock = NBaobab::GetAncestorResultBlock(surveyBlock)) {
            requestData["result_block_id"] = resultBlock->GetID();
            if (const auto& documentUrl = NBaobab::GetAttrValueUnsafe<TString>(*resultBlock, "documentUrl")) {
                requestData["url"] = documentUrl.GetRef();
            }
            if (const auto& wizardName = NBaobab::GetAttrValueUnsafe<TString>(*resultBlock, "wizard_name")) {
                requestData["wizard_name"] = wizardName.GetRef();
            }
            if (const auto& visible = CheckBlockVisible(request, joiner, resultBlock.GetRef())) {
                requestData["visibility"] = visible.GetRef();
            }
            if (const auto& wizardPos = NBaobab::GetAttrValueUnsafe<TString>(*resultBlock, "pos")) {
                requestData["wizard_position"] = wizardPos.GetRef();
            }
        }
        if (const auto& surveyId = NBaobab::GetAttrValueUnsafe<TString>(surveyBlock, "surveyId")) {
            requestData["survey_id"] = surveyId.GetRef();
        }
        requestData["answers"] = GetAnswers(surveyBlock);
        requestData["multi_answers"] = GetMultiAnswers(surveyBlock);
        requestData["survey_type"] = GetSurveyType(surveyBlock);
        if (const auto& questions = NBaobab::GetAttrValueUnsafe<NBaobab::TAttrsArray>(surveyBlock, "questions")) {
            requestData["questions"] = SerializeToNode(questions.GetRef());
        }
        if (const auto& prismSegment = GetPrismSegment(request)){
            requestData["prism_segment"] = prismSegment.GetRef();
        }
        requestData["dwelltime"] = request->GetDwellTime();

        return requestData;
    }

    NYT::TNode::TListType GetEventsData(
        const NYT::TNode& requestData,
        const NBaobab::TRequestJoiner& joiner,
        const NBaobab::TBlock& surveyBlock
    ) {
        NYT::TNode::TListType result;
        for (const auto& event : joiner.GetEventsByBlockSubtree(surveyBlock)) {
            auto eventData = requestData;
            eventData["ts"] = event->GetContext().GetServerTime();
            if (const auto clientEvent = dynamic_cast<const NBaobab::IClientEvent*>(event.Get())) {
                if (clientEvent->GetEventID().Defined()) {
                    eventData["event_id"] = clientEvent->GetEventID().GetRef();
                }
                eventData["block_id"] = clientEvent->GetBlockID();
            }

            if (const auto click = dynamic_cast<const NBaobab::IClick*>(event.Get())) {
                eventData["type"] = "click";

                if (auto clickBlock = NBaobab::GetBlockByID(joiner, click->GetBlockID())) {
                    eventData["block_name"] = clickBlock->GetName();
                    if (const auto& answerId = GetAnswerId(clickBlock.GetRef())) {
                        eventData["answer_id"] = answerId.GetRef();
                    }
                    if (const auto& idQuestion = NBaobab::GetAttrValueUnsafe<i64>(clickBlock.GetRef(), "idQuestion")) {
                        eventData["question_id"] = idQuestion.GetRef();
                    }
                }
                result.push_back(eventData);
            }
            else if (const auto tech = dynamic_cast<const NBaobab::ITech*>(event.Get())) {
                eventData["type"] = "techevent";

                if (const auto& commentText = NBaobab::GetAttrValueUnsafe<TString>(tech->GetData(), "commentText")) {
                    eventData["comment_text"] = commentText.GetRef();
                }
                if (const auto& selectedOption = NBaobab::GetAttrValueUnsafe<i64>(tech->GetData(), "selectedOption")) {
                    eventData["selected_option"] = selectedOption.GetRef();
                }
                if (const auto& commentId = NBaobab::GetAttrValueUnsafe<i64>(tech->GetData(), "commentId")) {
                    eventData["question_id"] = commentId.GetRef();
                }
                result.push_back(eventData);
            }
        }
        std::sort(result.begin(), result.end(), [](const NYT::TNode& e1, const NYT::TNode& e2) {
            return e1["ts"].AsInt64() < e2["ts"].AsInt64();
        });
        return result;
    }
}

void TWebSurveysActionsSqueezer::GetActions(TActionSqueezerArguments& args) const
{
    for (const NRA::TRequest* request : args.Container.GetRequests()) {
        if (!CheckRequest(request)) {
            continue;
        }
        try {
            if (auto baobabProps = dynamic_cast<const NRA::TBaobabProperties*>(request)) {
                const auto& joiner = baobabProps->GetRequestJoiner();
                const TMaybe<NBaobab::TBlock> root = joiner.GetRoot();
                if (root) {
                    auto bucketInfo = CheckExperiments(args, request);
                    for (const auto& surveyBlock : NBaobab::TBFSer(*root, NBaobab::NameFilter("user-survey"))) {
                        auto requestData = GetRequestData(request, joiner, surveyBlock);
                        args.ResultActions.emplace_back(requestData, bucketInfo);
                        for (const auto& eventData : GetEventsData(requestData, joiner, surveyBlock)) {
                            args.ResultActions.emplace_back(eventData, bucketInfo);
                        }
                    }
                }
            }
        } catch (...) {
            Cerr << "Fail a request process: reqid: " << request->GetReqID() << Endl;
        }
    }
}
/////////////////////////////// TWebSurveysActionsSqueezer End

}; // NMstand

File: tools/mstand/session_squeezer/squeezer_web_surveys.py
from collections import defaultdict

from typing import Tuple
from typing import Generator
from typing import List
from typing import Union
from yaqtypes import JsonDict

import session_squeezer.squeezer_common as sq_common

type_v3_list_tuple = {
    "type_name": "optional",
    "item": {
        "type_name": "list",
        "item": {
            "type_name": "tuple",
            "elements": [
                {
                    "type": "int32"
                },
                {
                    "type": {
                        "type_name": "list",
                        "item": "int32"
                    }
                }
            ]
        }
    }
}

WEB_SURVEYS_YT_SCHEMA = [
    {"name": "yuid", "type": "string"},
    {"name": "ts", "type": "int64"},
    {"name": "action_index", "type": "int64"},
    {"name": "bucket", "type": "int64"},
    {"name": "is_match", "type": "boolean"},
    {"name": "testid", "type": "string"},
    {"name": "servicetype", "type": "string"},

    {"name": "answer_id", "type": "int32"},
    {"name": "answers", "type": "any"},
    {"name": "multi_answers", "type": "any", "type_v3": type_v3_list_tuple},
    {"name": "block_id", "type": "string"},
    {"name": "block_name", "type": "string"},
    {"name": "comment_text", "type": "string"},
    {"name": "event_id", "type": "string"},
    {"name": "query_text", "type": "string"},
    {"name": "question_id", "type": "int32"},
    {"name": "questions", "type": "any"},
    {"name": "reqid", "type": "string"},
    {"name": "result_block_id", "type": "string"},
    {"name": "selected_option", "type": "int32"},
    {"name": "survey_id", "type": "string"},
    {"name": "survey_type", "type": "string"},
    {"name": "type", "type": "string"},
    {"name": "url", "type": "string"},
    {"name": "wizard_name", "type": "string"},
    {"name": "prism_segment", "type": "int64"},
    {"name": "dwelltime", "type": "int64"},
    {"name": "wizard_position", "type": "int64"},
]


class ActionsSqueezerWebSurveys(sq_common.ActionsSqueezer):
    """
    1: initial version
    2: add query_text field (MSTAND-2045)
    3: add answers order and survey type (MSTAND-2059)
    4: Fix the answers list formation (MSTAND-2059)
    5: add questions column (MSTAND-2075)
    6: add url column (MSTAND-2088)
    7: fix the answer list formation (MSTAND-2127)
    8: update for new answer format: answer block with id, label and idQuestion attrs (MSTAND-2128)
    9: add ui, browser and visibility columns (MSTAND-2152)
    10: save comment id into question_id column (MSTAND-2154)
    """
    VERSION = 11

    YT_SCHEMA = WEB_SURVEYS_YT_SCHEMA

    def __init__(self) -> None:
        super(ActionsSqueezerWebSurveys, self).__init__()

    def get_actions(self, args: sq_common.ActionSqueezerArguments) -> None:
        for request in args.container.GetRequests():
            if not self.check_request(request):
                continue

            exp_bucket = self.check_experiments(args, request)
            collect_survey_actions(args.result_actions, exp_bucket, request)

    def check_request(self, request: "libra.TRequest") -> bool:
        return any((
            request.IsA("TYandexWebRequest"),
            request.IsA("TTouchYandexWebRequest"),
            request.IsA("TPadYandexWebRequest"),
            request.IsA("TMobileAppYandexWebRequest"),
        ))


def collect_survey_actions(
        known_actions: List[JsonDict],
        exp_bucket: sq_common.ExpBucketInfo,
        request: "libra.TRequest",
) -> None:
    base_data = {
        "ts": request.Timestamp,
        "reqid": request.ReqId,
        "query_text": request.CorrectedQuery,
        sq_common.EXP_BUCKET_FIELD: exp_bucket,
    }

    import baobab
    for block, joiner, tree in get_survey_blocks(request):
        result_block = baobab.common.get_ancestor_result_block(block)
        result_block_id = result_block.id if result_block else None
        wizard_name = result_block.get_attr("wizard_name") if result_block else None
        url = result_block.get_attr("documentUrl") if result_block else None

        request_data = dict(
            base_data,
            survey_id=block.get_attr("surveyId"),
            type="request",
            wizard_name=wizard_name,
            result_block_id=result_block_id,
            answers=get_answers(block),
            multi_answers=get_multi_answers(block),
            survey_type=get_survey_type(block),
            questions=block.get_attr("questions"),
            url=url,
        )

        known_actions.append(request_data)
        collect_events_data(known_actions, request_data, block, joiner, tree)


def get_answers(block: "baobab.common.Block") -> List[int]:
    import baobab
    name_filter = baobab.common.NameFilter("answer")
    return [
        get_answer_id(answer)
        for answer in baobab.common.bfs_iterator_filtered(block, name_filter)
    ]


def get_multi_answers(block: "baobab.common.Block") -> List[Tuple[int, List[int]]]:
    import baobab
    name_filter = baobab.common.NameFilter("answer")
    answers = defaultdict(list)
    for answer in baobab.common.bfs_iterator_filtered(block, name_filter):
        if "idQuestion" in answer.attrs:
            answers[answer.get_attr("idQuestion")].append(get_answer_id(answer))
    return list(answers.items())


def get_survey_type(block: "baobab.common.Block") -> str:
    import baobab
    name_filter = baobab.common.NameFilter("answer")
    for answer in baobab.common.bfs_iterator_filtered(block, name_filter):
        if "text" in answer.attrs:
            return "text"
        if "emoji_id" in answer.attrs:
            return "emoji"
    return "unknown"


def get_survey_blocks(
        request: "libra.TRequest",
) -> Generator[Union["baobab.common.Block", "baobab.common.ShowAndClicksJoiner", "baobab.common.Tree"], None, None]:
    import baobab

    if request.IsA("TBaobabProperties"):
        joiners = request.BaobabAllTrees()
        for joiner in joiners:
            tree = joiner.get_show().tree
            name_filter = baobab.common.NameFilter("user-survey")
            for block in baobab.common.bfs_iterator_filtered(tree.root, name_filter):
                yield block, joiner, tree


def get_answer_id(answer_block: "baobab.common.Block") -> str:
    return answer_block.get_attr("answer_id") if "answer_id" in answer_block.attrs else answer_block.get_attr("id")


def collect_events_data(
        known_actions: List[JsonDict],
        request_data: JsonDict,
        block: "baobab.common.Block",
        joiner: "baobab.common.ShowAndClicksJoiner",
        tree: "baobab.common.Tree",
) -> None:
    import baobab

    events_data = []
    for event in joiner.get_events_by_block_subtree(block):
        event_data = dict(
            request_data,
            ts=event.client_timestamp,
            event_id=event.event_id,
            block_id=event.block_id,
        )
        if isinstance(event, baobab.common.Tech):
            event_data["type"] = "techevent"
            event_data["comment_text"] = event.data.get("commentText")
            event_data["selected_option"] = event.data.get("selectedOption")
        elif isinstance(event, baobab.common.Click):
            click_block = tree.get_block_by_id(event.block_id)
            event_data["type"] = "click"
            event_data["block_name"] = click_block.name
            event_data["answer_id"] = get_answer_id(click_block)
            event_data["question_id"] = click_block.get_attr("idQuestion")

        events_data.append(event_data)

    events_data.sort(key=lambda x: x["ts"])
    known_actions.extend(events_data)
